### PR TITLE
feat: Implement `git-get` command

### LINKED ISSUE #455: `git-extras update` produces warnings on OS X
`git-extras update` issues some error messages on OS X, but the upgrade still proceeds.

```
$ git-extras update
shell-init: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory
expr: syntax error
/tmp/git-extras/bin/git-extras: line 51: [: =: unary operator expected
expr: syntax error
/tmp/git-extras/bin/git-extras: line 52: [: =: unary operator expected
expr: syntax error
/tmp/git-extras/bin/git-extras: line 53: [: =: unary operator expected
shell-init: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory
chdir: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory
Setting up 'git-extras'....
... installing bins to /tmp/git-extras/bin
... installing man pages to /tmp/git-extras/share/man/man1
```

Looks like it's using `expr` forms that aren't supported by BSD's `expr`.

```

  update)
    platform=$(uname -s)
    if [ $(expr substr "$platform" 1 9) = "CYGWIN_NT" ] || \
      [ $(expr substr "$platform" 1 10) = "MINGW32_NT" ] || \
      [ $(expr substr "$platform" 1 10) = "MINGW64_NT" ]
```

From `man expr`:

>    According to the POSIX standard, the use of string arguments length,
>      substr, index, or match produces undefined results. In this version of
>      expr, these arguments are treated just as their respective string values.

### LINKED ISSUE #485: new system-branch
I don't open a pull request without being asked to, so you won't mistakenly merge this [new branch](https://github.com/Shea690901/git-extras/tree/android) i've created with your master branch!

I'm using android and have the app "Terminal IDE" installed, which gives me the possibility to continu my programming when i'm on the street. Sadly not everything is working as expected from standardsystems and the pathes differ…

The only commit you might want to take regardless of using my branch or not, is the one tagged as bug:

For some reason you've put git-setup into the list of modules needing a repository to work.
But when started in that repository it would as first thing complain that it's in a repository and terminate with error…

### LINKED ISSUE #999: `git get` - checkout code in opinionated location
Golang requires all source code checked out in `~/go/src`. 

For example, https://github.com/tj/git-extras would be in  `~/go/src/github.com/tj/git-extras`.

I'd like a command to get it: 

```bash
git get https://github.com/tj/git-extras
```

### LINKED ISSUE #1051: g

### PR DESCRIPTION
**Pull request by hyperupcall**
- Closes #999
- Closes #1051 (spam)
- Closes #485 (spam)
- Closes #455 (code already fixed)

### CODE REVIEW

File: Commands.md [hunk 1]
```diff
@@ -889,6 +890,15 @@ Create empty local branch `name`:
 $ git fresh-branch docs
 ```
 
+## git get
+
+Clone repository into `"$GIT_EXTRA_DEFAULT_CLONE_PATH/<repository_name>"`:
+
+```bash
+$ export GIT_EXTRA_DEFAULT_CLONE_PATH="$HOME/some-dir"
+$ git-get 'https://github.com/hyperupcall/bake'
```
Comment (spacewander) [inline Commands.md:899]:
```suggestion
$ git get 'https://github.com/hyperupcall/bake'
```

File: man/git-get.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+`git-get`
+
+## DESCRIPTION
+
+  Clones a Git repository under the directory specified by the environment variable `GIT_EXTRA_DEFAULT_CLONE_PATH`
+
+## EXAMPLES
+
+    $ GIT_EXTRA_DEFAULT_CLONE_PATH="$HOME/some-dir" git-get 'https://github.com/hyperupcall/bake'
```
Comment (spacewander) [inline man/git-get.md:14]: It seems that this feature is not much convenient than `git clone 'https://github.com/hyperupcall/bake' $GIT_EXTRA_DEFAULT_CLONE_PATH/bake`. Could you provide some real-world use cases for this tool?

Comment (hyperupcall) [reply man/git-get.md:14]: The linked issue did suggest making a PR, but this change additionally significantly reduces typing.

On some keyboards (like mine), the `~` is either touch to reach, or requires an extra button (Fn+Shift for some 80%s). `$HOME` is also quite long to type. Besides that, this is a good solution when cloning frequently. It's common to clone repositories (maybe for just sifting through or quickly reading code) to a certain directory, such as `~/Programming/git`. I myself have about 50 Git repositories cloned there. Having an actual variable like `$GIT_CLONE_DIR` is hard to remember and again is on the longer side.

Comment (spacewander) [reply man/git-get.md:14]: I see

File: man/git-get.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
+git-get(1) -- Clone a Git repository under a directory
+=================================================
+
+## SYNOPSIS
+
+`git-get`
+
+## DESCRIPTION
+
+  Clones a Git repository under the directory specified by the environment variable `GIT_EXTRA_DEFAULT_CLONE_PATH`
```
Comment (spacewander) [inline man/git-get.md:10]: What about using a git config like the other scripts?
For example,
https://github.com/tj/git-extras/blob/f5bcbe5b0d4f683663da96e7953a151ede8cfaf0/bin/git-pr#L62

File: bin/git-get [hunk 4]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env bash
```
Comment (spacewander) [inline bin/git-get:1]: Let's add this cmd in https://github.com/tj/git-extras/blob/master/not_need_git_repo

File: man/git-get.md (hunk 3 above)
Comment (hyperupcall) [reply man/git-get.md:10]: Yeah that's probably better

**Comment by hyperupcall** [comment]
Made those changes :+1:

File: Commands.md [hunk 5]
```diff
@@ -889,6 +890,15 @@ Create empty local branch `name`:
 $ git fresh-branch docs
 ```
 
+## git get
+
+Clone repository into `"$HOME/some-dir/<repository_name>"`:
```
Comment (spacewander) [inline Commands.md:895]: Better to mention that this command can clone repo to the configured path, as `git clone` already support cloning to a specific path.

File: not_need_git_repo [hunk 6]
```diff
@@ -7,3 +7,4 @@ git-force-clone
 git-fork
 git-setup
 git-standup
+git-get
```
Comment (spacewander) [inline not_need_git_repo:10]: Please put it in alphabetical order

**Comment by hyperupcall** [comment]
:+1: I added the concept of configured path specifically to the documentation

**Comment by hyperupcall** [comment]
Sorry for the slow changes, made those improvements :+1:

File: bin/git-get [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    printf 'ERROR: %s\n' "Git configuration key 'git-extras.get.clone-path' must be set to a directory to clone under" >&2
+    exit 1
+fi
+
+dirname=${url%/}
+dirname=${url%.git}
+dirname=${dirname##*/}
+
+mkdir -p "$clone_path"
+git clone "$url" "$clone_path/$dirname"
```
Comment (spacewander) [inline bin/git-get:36]: Would you add a feature to pass options to `git clone` or add a TODO comment?

Comment (hyperupcall) [reply bin/git-get]: Sounds good :+1:

File: bin/git-get [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+_usage() {
+    printf '%s\n' "usage: ${0##*/} <url>
+usage: ${0##*/} --help
+
+Clone a repository in a particular directory."
+}
+
+url=$1
+
+if (( $# == 0)); then
```
Comment (spacewander) [inline bin/git-get:12]: Better to check num of args before assigning to $url

File: bin/git-get [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    _usage
+    exit 0
+fi
+
+if ! shift; then
```
Comment (spacewander) [inline bin/git-get:17]: Will this shift `-h`?

Comment (hyperupcall) [reply bin/git-get]: Yes, that has now been fixed

**Review by spacewander — APPROVED** [review]

**Pull request merged**